backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -129,64 +111,43 @@ private void previewPinOfPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<
-
-        for (int socket : getOnlineSockets(cpuTopology)) {
-            int coresInSockets = getCoresInSocket(cpuTopology, socket).size();
-            if (hostCoresPerSocket * hostThreadsPerCore == coresInSockets &&
-                    coresInSockets >= vcpus &&
-                    getFreeCpusInSocket(cpuTopology, socket).size() < vcpus) {
-                // we should fit into one socket
-                continue;
-            }
-            cpusLeft = allocateCores(cpuTopology, hostThreadsPerCore, socket, cpusLeft,
-                    cpusToBeAllocated, true, vm.getId(), vm.getCpuPinningPolicy());
-            if (cpusLeft > 0) {
-                // iterate again on the cores, take whatever we can.
-                cpusLeft = allocateCores(cpuTopology, hostThreadsPerCore, socket, cpusLeft,
-                        cpusToBeAllocated, false, vm.getId(), vm.getCpuPinningPolicy());
-            }
-            if (cpusLeft == 0) {
-                return cpusToBeAllocated;
-            }
-        }
-        return cpusLeft == 0 ? cpusToBeAllocated : new ArrayList<>();
+        cpusToBeAllocated = allocateCpus(cpuTopology, vm);
+        return cpusToBeAllocated.size() == vm.getNumOfCpus() ? cpusToBeAllocated : null;
     }
 
-    private int allocateCores(List<VdsCpuUnit> cpuTopology, int hostThreadsPerCore, int socket, int cpusLeft,
-                              List<VdsCpuUnit> cpusToBeAllocated, boolean wholeCore, Guid vmId,
-                              CpuPinningPolicy cpuPinningPolicy) {
-        for (int core : getOnlineCores(cpuTopology, socket)) {
-            List<VdsCpuUnit> sharedCpus = getFreeCpusInCore(cpuTopology, socket, core);
-            int sharedCpusInCore = sharedCpus.size();
-            List<Integer> sharedCpuIds = sharedCpus.stream().map(VdsCpuUnit::getCpu).collect(Collectors.toCollection(LinkedList::new));
-            Iterator<Integer> sharedCpusIterator = sharedCpuIds.iterator();
-            if (cpusLeft == 0) {
-                return cpusLeft;
-            }
-            if (wholeCore) {
-                if (cpusLeft >= hostThreadsPerCore && sharedCpusInCore == hostThreadsPerCore) {
-                    // take the whole core
-                    while (sharedCpusInCore-- > 0) {
-                        VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                        cpusToBeAllocated.add(cpuToTake);
-                        cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                        cpusLeft--;
+    private List<VdsCpuUnit> allocateCpus(List<VdsCpuUnit> cpuTopology, VM vm) {
+        List<VdsCpuUnit> cpusToBeAllocated = new ArrayList<>();
+        int socketsLeft = vm.getNumOfSockets();
+        int onlineSockets = getOnlineSockets(cpuTopology).size();
+        while (onlineSockets > 0 && socketsLeft > 0) {
+            List<VdsCpuUnit> cpusInChosenSocket = getMaxFreedSocket(cpuTopology);
```

Comment:
would it make sense to take threadPerCore into account when looking for the max freed socket?

Reply:
We talked about it offline and this comment is irrelevant, right?

Reply:
right, how about adding a comment to this method that reminds us we can assume that a valid pinning exists here (because the host was filtered beforehand) to prevent such confusing